akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala

```diff
@@ -1140,8 +1153,8 @@ private[cluster] class OnMemberUpListener(callback: Runnable) extends Actor with
     }
   }
 
-  def isSelfUp(m: Member): Boolean =
-    m.uniqueAddress == cluster.selfUniqueAddress && m.status == MemberStatus.Up
+  private def checkChanged(m: Member, previousStatus: MemberStatus = null): Boolean =
+    m.uniqueAddress == cluster.selfUniqueAddress && (if (prefer == Down) previousStatus == prefer else m.status == prefer)
```

Comment:
Hmm, programmatic actions like the ones you cite are not the only ways to join or leave a cluster, therefore I’d rather have the generic hooks than future-returning actions. It is pretty trivial to write an extension that then uses these hooks to provide one Future for each purpose.